fdbserver/BlobManager.actor.cpp

```diff
@@ -822,6 +832,146 @@ ACTOR Future<Void> rangeMover(BlobManagerData* bmData) {
+int numExistingBWOnAddr(BlobManagerData* self, const AddressExclusion& addr) {
+	int numExistingBW = 0;
+	for (auto& server : self->workersById) {
+		const NetworkAddress& netAddr = server.second.stableAddress();
+		AddressExclusion usedAddr(netAddr.ip, netAddr.port);
+		if (usedAddr == addr) {
+			++numExistingBW;
+		}
+	}
+
+	return numExistingBW;
+}
+
+// Tries to recruit a blob worker on the candidateWorker process
+ACTOR Future<Void> initializeBlobWorker(BlobManagerData* self, RecruitBlobWorkerReply candidateWorker) {
+	const NetworkAddress& netAddr = candidateWorker.worker.stableAddress();
+	AddressExclusion workerAddr(netAddr.ip, netAddr.port);
+	// Ask the candidateWorker to initialize a BW only if the worker does not have a pending request
+	if (numExistingBWOnAddr(self, workerAddr) == 0 &&
+	    self->recruitingLocalities.count(candidateWorker.worker.stableAddress()) == 0) {
+		state UID interfaceId = deterministicRandom()->randomUniqueID();
+
+		state InitializeBlobWorkerRequest initReq;
+		initReq.reqId = deterministicRandom()->randomUniqueID();
+		initReq.interfaceId = interfaceId;
+
+		// acknowledge that this worker is currently being recruited on
+		self->recruitingLocalities.insert(candidateWorker.worker.stableAddress());
+
+		TraceEvent("BMRecruiting")
+		    .detail("State", "Sending request to worker")
+		    .detail("WorkerID", candidateWorker.worker.id())
+		    .detail("WorkerLocality", candidateWorker.worker.locality.toString())
+		    .detail("Interf", interfaceId)
+		    .detail("Addr", candidateWorker.worker.address());
+
+		// send initialization request to worker (i.e. worker.actor.cpp)
+		// here, the worker will construct the blob worker at which point the BW will start!
+		Future<ErrorOr<InitializeBlobWorkerReply>> fRecruit =
+		    candidateWorker.worker.blobWorker.tryGetReply(initReq, TaskPriority::BlobManager);
+
+		// wait on the reply to the request
+		state ErrorOr<InitializeBlobWorkerReply> newBlobWorker = wait(fRecruit);
+
+		// if the initialization failed in an unexpected way, then kill the BM.
+		// if it failed in an expected way, add some delay before we try to recruit again
+		// on this worker
+		if (newBlobWorker.isError()) {
+			TraceEvent(SevWarn, "BMRecruitmentError").error(newBlobWorker.getError());
+			if (!newBlobWorker.isError(error_code_recruitment_failed) &&
+			    !newBlobWorker.isError(error_code_request_maybe_delivered)) {
+				throw newBlobWorker.getError();
+			}
+			wait(delay(SERVER_KNOBS->STORAGE_RECRUITMENT_DELAY, TaskPriority::BlobManager));
+		}
+
+		// if the initialization succeeded, add the blob worker's interface to
+		// the blob manager's data and start monitoring the blob worker
+		if (newBlobWorker.present()) {
+			BlobWorkerInterface bwi = newBlobWorker.get().interf;
+
+			self->workersById.insert({ bwi.id(), bwi });
+			self->workerStats.insert({ bwi.id(), BlobWorkerStats() });
+			self->addActor.send(monitorBlobWorker(self, bwi));
```

Comment:
Not specific to this line, but we might not want to use the same `ActorCollection` across the entire blob manager. The reason being is that if we decide to throw in one of the Actors that get added to the collection, then we'll end up killing the blob manager, which might not be ideal?

Reply:
I think we do want to use a global actor collection or error forwarder, and like already done in this function, any "expected" errors should not be thrown up to the top to kill the worker.
Any "unexpected" errors should kill the worker.

Reply:
Sounds good 👍